### ISSUE TITLE
Wakanda server dataclass sum is broken

### ISSUE THREAD

**Issue by vegarringdal** [Wakanda/wakanda-issues#246]
**Wakanda studio/server:**
- Wakanda Studio  2.4.0
**OS**
- Windows 10, 64bit, vmware

**Model:**
![image](https://user-images.githubusercontent.com/2901416/37482638-0717db78-2885-11e8-8011-457c3b3d4eba.png)

```json
{
	"toJSON": true,
	"extraProperties": {
		"version": "2",
		"classes": {
			"DataClass1": {
				"panelColor": "#76923C",
				"panel": {
					"isOpen": "true",
					"position": {
						"X": 25,
						"Y": 20
					},
					"pathVisible": true
				}
			}
		},
		"model": {
			"scriptMode": "manual",
			"workspaceLeft": 0,
			"workspaceTop": 0
		}
	},
	"dataClasses": [
		{
			"collectionName": "DataClass1Collection",
			"className": "DataClass1",
			"attributes": [
				{
					"name": "ID",
					"kind": "storage",
					"type": "long",
					"autosequence": "true",
					"unique": "true",
					"primKey": "true"
				},
				{
					"name": "person",
					"scope": "public",
					"type": "string",
					"kind": "storage"
				},
				{
					"name": "salery",
					"scope": "public",
					"type": "number",
					"kind": "storage"
				}
			]
		}
	],
	"type": []
}
```

**Code:**

```js
ds.DataClass1.all().remove();

var x = new ds.DataClass1();
x.person = "test1";
x.salery = 0.1;
x.save();

var y = new ds.DataClass1();
y.person = "test2";
y.salery = 0.2;
y.save();



ds.DataClass1.all().sum('salery');
```

**Error:**
![image](https://user-images.githubusercontent.com/2901416/37482679-212fe35c-2885-11e8-91b6-8df501c08a33.png)

Why isnt this 0.3 only ? I would epect to get correct value when database did the summary.

According to docs the number attibute is not a float
http://doc.wakanda.org/home2.en.html#/Data-Model-Editor/Attributes.300-574626.en.html
![image](https://user-images.githubusercontent.com/2901416/37482613-f58b16d6-2884-11e8-9c5e-c331807fe1e9.png)

Or Im I missing something ?

**Comment by erlicthemad** [comment]
You are actually seeing a fun bit of Javascript entertainment. It is related to the way Javascript stores and handles numbers.  In Javascript, a number is its a decimal wind up being of a type where you get approximations. As a result, your sum is doing what it should. You will either need to round off the number or store your number shifted by two decimal places. Save your .3 as 30, then have a number 1 be 100. This makes the number integers in the system and you get rid of the approximations. Or when you do math shift two decimal places, do the math then shift them back when returning the result.

**Comment by vegarringdal** [comment]
@erlicthemad 

What ?
Im doing sum in the database on two records, not js..( I know this will happend in js due to floating point) database should be able to sum to records/entities with values  0.1 + 0.3.

**Comment by erlicthemad** [comment]
Yes Wakanda is using javascript for its math operations. This is explained in the following link page.

https://www.w3schools.com/js/js_numbers.asp <https://www.w3schools.com/js/js_numbers.asp>

This is a clip from the examples.

var x = 0.2 + 0.1;         // x will be 0.30000000000000004

**Comment by vegarringdal** [comment]
@erlicthemad 

thx for replying, I know how javascript part work, but still expect more from database sum() function to be able to handle this. **±1.7e±308 (real)** just expected this to managed this better, but no expert in floating points/double precision in any way. 

If I do similar in postgresI dont get result like this.
**Table content:**
![image](https://user-images.githubusercontent.com/2901416/37491513-6d7b924c-289e-11e8-8c39-59bee09e21d9.png)

**Sum result**
![image](https://user-images.githubusercontent.com/2901416/37491552-8a304eb4-289e-11e8-97c3-37050b916f5e.png)

**Comment by erlicthemad** [comment]
Yes, Sadly your right. One would expect that a database would add up numbers right. I would expect that it would do this too. But Wakanda in written primarily in Javascript (V8 to be precise). As a result its DB engine is subject to the same issues as any other javascript function. You can always try to report it as a bug. But since its the way Javascript works I would not expect them to fix it. When you perform a sum function all they do is execute the a Javascript summing function.

Whit your expecting is not beyond reason, but in Wakanda they do all of it in Javascript. Offhand a fix might be to use a real DB storage engine and let the engine do the math. Otherwise you would have to write the db functions as you would write in Javascript. 

I am not a Wakanda staffer, I’m just a user who had beat my own head against this bug before. Lets see if they acknowledge it and fix it. Or they say its how Javascript works and get used to it.

**Comment by vegarringdal** [comment]
@erlicthemad 

The database engine/core is not js, thats c or c++, now with Google V8 engine for running the js parts.

**Comment by hamzahik** [comment]
@vegarringdal @erlicthemad I'll discuss this with the team and get back to you.

**Comment by vegarringdal** [comment]
@hamzahik 
Did anything happen here ?
